pkg/metrics/report.go

```diff
@@ -64,14 +69,45 @@ func ReportResourceUsage(ctx context.Context, kubeClient *kubernetes.Clientset,
 				}
 			}
 			// get resource limits
-			if v, ok := limits[corev1.ResourceName(resource)]; ok {
+			if v, ok := list.limits[corev1.ResourceName(resource)]; ok {
 				val = v.AsApproximateFloat64()
 			} else {
 				val = 0
 			}
 			nodeResourceLimits.WithLabelValues(labels...).Set(val)
 		}
 	}
+	log.V(4).Infof("Reporting cycle took %s", time.Since(start).String())
+}
+
+func getNodeResourceMap(ctx context.Context, kubeClient *kubernetes.Clientset, namespace string) (map[string]*resourceList, error) {
+	pods, err := kubeClient.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list the pods: %v", err)
+	}
+	log.V(4).Infof("Found %d pods in %q namespace", len(pods.Items), namespace)
+
+	nodeResources := make(map[string]*resourceList)
+	for _, pod := range pods.Items {
```

Comment:
Can we parallelize the for loop? In many cases for large multi-node training pods will have a 1:1 mapping to nodes